File: library/src/test/java/io/ashdavies/rx/rxfirebase/RxFirebaseAuthTest.java
package io.ashdavies.rx.rxfirebase;

import com.google.firebase.auth.AuthCredential;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.runners.MockitoJUnitRunner;

import static com.google.common.truth.Truth.assertThat;
import static org.mockito.BDDMockito.given;
import static org.mockito.BDDMockito.then;
import static org.mockito.Matchers.any;
import static org.mockito.Mockito.mock;

@RunWith(MockitoJUnitRunner.class)
public class RxFirebaseAuthTest {

  private RxFirebaseAuth rx;

  @Mock FirebaseAuth auth;

  @Before
  public void setUp() throws Exception {
    rx = RxFirebaseAuth.getInstance(auth);
  }

  @Test
  public void shouldAddAuthStateListener() throws Exception {
    rx.authStateChanged().subscribe();

    then(auth).should().addAuthStateListener(any(FirebaseAuth.AuthStateListener.class));
  }

  @Test
  public void shouldSignInWithCredentials() throws Exception {
    AuthCredential credential = mock(AuthCredential.class);

    rx.signInWithCredential(credential).subscribe();

    then(auth).should().signInWithCredential(credential);
  }

  @Test
  public void shouldSignInWithCustomToken() throws Exception {
    rx.signInWithCustomToken("token").subscribe();

    then(auth).should().signInWithCustomToken("token");
  }

  @Test
  public void shouldSignInWithEmailAndPassword() throws Exception {
    rx.signInWithEmailAndPassword("email", "password").subscribe();

    then(auth).should().signInWithEmailAndPassword("email", "password");
  }

  @Test
  public void shouldSignInAnonymously() throws Exception {
    rx.signInAnonymously().subscribe();

    then(auth).should().signInAnonymously();
  }

  @Test
  public void shouldCreateUserWithEmailAndPassword() throws Exception {
    rx.createUserWithEmailAndPassword("email", "password").subscribe();

    then(auth).should().createUserWithEmailAndPassword("email", "password");
  }

  @Test
  public void shouldFetchProvidersForEmail() throws Exception {
    rx.fetchProvidersForEmail("email").subscribe();

    then(auth).should().fetchProvidersForEmail("email");
  }

  @Test
  public void shouldGetCurrentUser() throws Exception {
    given(auth.getCurrentUser()).willReturn(mock(FirebaseUser.class));

    rx.getCurrentUser().test()
        .assertValueCount(1)
        .assertComplete();
  }

  @Test
  public void shouldSendPasswordResetEmail() throws Exception {
    rx.sendPasswordResetEmail("email").subscribe();

    then(auth).should().sendPasswordResetEmail("email");
  }

  @Test
  public void shouldSignOut() throws Exception {
    assertThat(rx.signOut()).isEqualTo(rx);

    then(auth).should().signOut();
  }

  @Test
  public void shouldNotGetCurrentUser() throws Exception {
    rx.getCurrentUser().test()
        .assertNoValues()
        .assertComplete();
  }
}

File: src/test/kotlin/io/ashdavies/rx/rxfirebase/RxFirebaseAuthTest.kt
package io.ashdavies.rx.rxfirebase

import com.google.android.gms.tasks.Tasks
import com.google.firebase.auth.AuthCredential
import com.google.firebase.auth.AuthResult
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.auth.ProviderQueryResult
import com.google.firebase.auth.SignInMethodQueryResult
import com.nhaarman.mockito_kotlin.given
import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.then
import org.junit.jupiter.api.Test

internal class RxFirebaseAuthTest {

  private val result: AuthResult = mock()
  private val delegate: FirebaseAuth = mock()

  private val auth = RxFirebaseAuth(delegate)

  @Test
  fun `should return current user`() {
    val user: FirebaseUser = mock()

    given(delegate.currentUser).willReturn(user)

    auth
        .currentUser()
        .test()
        .assertResult(user)
  }

  @Test
  fun `should sign in with credential`() {
    val credential: AuthCredential = mock()

    given(delegate.signInWithCredential(credential)).willReturn(Tasks.forResult(result))

    auth
        .signInWithCredential(credential)
        .test()
        .assertResult(result)
  }

  @Test
  fun `should sign in with custom token`() {
    given(delegate.signInWithCustomToken("CUSTOM_TOKEN")).willReturn(Tasks.forResult(result))

    auth
        .signInWithCustomToken("CUSTOM_TOKEN")
        .test()
        .assertResult(result)
  }

  @Test
  fun `should sign in with email and password`() {
    given(delegate.signInWithEmailAndPassword("EMAIL", "PASSWORD")).willReturn(Tasks.forResult(result))

    auth
        .signInWithEmailAndPassword("EMAIL", "PASSWORD")
        .test()
        .assertResult(result)
  }

  @Test
  fun `should sign in anonymously`() {
    given(delegate.signInAnonymously()).willReturn(Tasks.forResult(result))

    auth
        .signInAnonymously()
        .test()
        .assertResult(result)
  }

  @Test
  fun `should create user with email and password`() {
    given(delegate.createUserWithEmailAndPassword("EMAIL", "PASSWORD")).willReturn(Tasks.forResult(result))

    auth
        .createUserWithEmailAndPassword("EMAIL", "PASSWORD")
        .test()
        .assertResult(result)
  }

  @Test
  fun `should fetch providers for email`() {
    val result: ProviderQueryResult = mock()

    given(delegate.fetchProvidersForEmail("EMAIL")).willReturn(Tasks.forResult(result))

    auth
        .fetchProvidersForEmail("EMAIL")
        .test()
        .assertResult(result)
  }

  @Test
  fun `should fetch sign in methods for email`() {
    val result: SignInMethodQueryResult = mock()

    given(delegate.fetchSignInMethodsForEmail("EMAIL")).willReturn(Tasks.forResult(result))

    auth
        .fetchSignInMethodsForEmail("EMAIL")
        .test()
        .assertResult(result)
  }

  @Test
  fun `should send password reset email`() {
    given(delegate.sendPasswordResetEmail("EMAIL")).willReturn(Tasks.forResult(null))

    auth
        .sendPasswordResetEmail("EMAIL")
        .test()
        .assertResult()
  }

  @Test
  fun `should sign out user`() {
    auth.signOut()

    then(delegate)
        .should()
        .signOut()
  }
}
